node/core/pvf/src/host.rs

```diff
@@ -0,0 +1,1049 @@
+				prepare_queue,
+				prepare::ToQueue::Enqueue {
+					priority: Priority::Background,
+					pvf: active_pvf,
+				},
+			)
+			.await?;
+		}
+	}
+
+	Ok(())
+}
+
+async fn handle_prepare_done(
+	cache_path: &Path,
+	artifacts: &mut Artifacts,
+	execute_queue: &mut mpsc::Sender<execute::ToQueue>,
+	awaiting_prepare: &mut AwaitingPrepare,
+	artifact_id: ArtifactId,
+) -> Result<(), Fatal> {
+	// Make some sanity checks and extract the current state.
+	let state = match artifacts.artifact_state_mut(&artifact_id) {
+		None => {
+			// before sending request to prepare, the artifact is inserted with `preparing` state;
+			// the requests are deduplicated for the same artifact id;
+			// there is only one possible state change: prepare is done;
+			// thus the artifact cannot be unknown, only preparing;
+			// qed.
+			never!("an unknown artifact was prepared: {:?}", artifact_id);
+			return Ok(());
+		}
+		Some(ArtifactState::Prepared { .. }) => {
+			// before sending request to prepare, the artifact is inserted with `preparing` state;
+			// the requests are deduplicated for the same artifact id;
+			// there is only one possible state change: prepare is done;
+			// thus the artifact cannot be prepared, only preparing;
+			// qed.
+			never!("the artifact is already prepared: {:?}", artifact_id);
+			return Ok(());
+		}
+		Some(state @ ArtifactState::Preparing) => state,
+	};
+
+	// It's finally time to dispatch all the execution requests that were waiting for this artifact
+	// to be prepared.
+	let artifact_path = artifact_id.path(&cache_path);
+	let pending_requests = awaiting_prepare.take(&artifact_id);
+	for PendingExecutionRequest { params, result_tx } in pending_requests {
+		if result_tx.is_canceled() {
+			// Preparation could've taken quite a bit of time and the requester may be not interested
+			// in execution anymore, in which case we just skip the request.
+			continue;
+		}
+
+		send_execute(
+			execute_queue,
+			execute::ToQueue::Enqueue {
+				artifact_path: artifact_path.clone(),
+				params,
+				result_tx,
+			},
+		)
+		.await?;
+	}
+
+	// Now consider the artifact prepared.
+	*state = ArtifactState::Prepared {
+		last_time_needed: SystemTime::now(),
+	};
+
+	Ok(())
+}
+
+async fn send_prepare(
+	prepare_queue: &mut mpsc::Sender<prepare::ToQueue>,
+	to_queue: prepare::ToQueue,
+) -> Result<(), Fatal> {
+	prepare_queue.send(to_queue).await.map_err(|_| Fatal)
+}
+
+async fn send_execute(
+	execute_queue: &mut mpsc::Sender<execute::ToQueue>,
+	to_queue: execute::ToQueue,
+) -> Result<(), Fatal> {
+	execute_queue.send(to_queue).await.map_err(|_| Fatal)
+}
+
+async fn handle_cleanup_pulse(
+	cache_path: &Path,
+	sweeper_tx: &mut mpsc::Sender<PathBuf>,
+	artifacts: &mut Artifacts,
+	artifact_ttl: Duration,
+) -> Result<(), Fatal> {
+	let to_remove = artifacts.prune(artifact_ttl);
+	for artifact_id in to_remove {
+		let artifact_path = artifact_id.path(cache_path);
+		sweeper_tx.send(artifact_path).await.map_err(|_| Fatal)?;
```

Comment:
the error is unreachable because `sweeper_task` only stops when the sender is dropped

Reply:
Yeah, and you can say the same about the other uses of `Fatal`, with the exception that here it is trivially visible. Not sure if expect would be more appropriate here.

Reply:
probably fine as is, just an observation